lib/ansible/modules/cloud/docker/docker_node.py

```diff
@@ -0,0 +1,290 @@
+       install the C(docker) Python module. Note that both modules should I(not)
+       be installed at the same time. Also note that when both modules are installed
+       and one of them is uninstalled, the other might no longer function and a
+       reinstall of it is required."
+    - "The docker server >= 1.10.0"
+author:
+  - Piotr Wojciechowski (@wojciechowskipiotr)
+  - Thierry Bouvet (@tbouvet)
+
+'''
+
+EXAMPLES = '''
+- name: List names of registered swarm nodes
+  docker_node:
+    state: list
+
+- name: Set node role
+  docker_node:
+    state: update
+    name: mynode
+    role: manager
+
+- name: Set node availability
+  docker_node:
+    state: update
+    name: mynode
+    availability: drain
+
+- name: Replace node labels with new labels
+  docker_node:
+    state: update
+    name: mynode
+    labels:
+      key: value
+    labels_state: replace
+
+- name: Merge node labels and new labels
+  docker_node:
+    state: update
+    name: mynode
+    labels:
+      key: value
+
+- name: Merge node labels and new labels
+  docker_node:
+    state: update
+    name: mynode
+    labels:
+      key: value
+    labels_state: merge
+
+- name: Remove node labels
+  docker_node:
+    state: update
+    name: mynode
+    labels_state: remove
+'''
+
+RETURN = '''
+node_facts:
+  description: Information about node after 'update' operation
+  returned: success
+  type: dict
+actions:
+  description: Provides the actions done on the swarm.
+  returned: when action failed.
+  type: list
+  example: "['This cluster is already a swarm cluster']"
+
+'''
+
+try:
+    from docker.errors import APIError
+except ImportError:
+    # missing docker-py handled in ansible.module_utils.docker_common
+    pass
+
+from ansible.module_utils.docker_common import (
+    DockerBaseClass,
+)
+
+from ansible.module_utils._text import to_native
+
+from ansible.module_utils.docker_swarm import AnsibleDockerSwarmClient
+
+
+class TaskParameters(DockerBaseClass):
+    def __init__(self, client):
+        super(TaskParameters, self).__init__()
+
+        self.state = None
+
+        # Spec
+        self.name = None
+        self.labels = None
+        self.labels_state = None
+
+        # Node
+        self.availability = None
+        self.role = None
+
+        for key, value in client.module.params.items():
+            setattr(self, key, value)
+
+
+class SwarmNodeManager(DockerBaseClass):
+
+    def __init__(self, client, results):
+
+        super(SwarmNodeManager, self).__init__()
+
+        self.client = client
+        self.results = results
+        self.check_mode = self.client.check_mode
+
+        self.parameters = TaskParameters(client)
+
+    def __call__(self):
+        choice_map = {
+            "update": self.node_update,
+        }
+
+        choice_map.get(self.parameters.state)()
+
+    def node_update(self):
+        if not (self.client.check_if_swarm_manager()):
```

Comment:
Why not move this (and the next) check to the constructor?

Reply:
That can be done, but needs to include an exception for `state: join` in `docker_swarm` module. I understand `docker_swarm` will use the `AnsibleDockerSwarmClient` class at some point, so we can add the required exception now or later.

Reply:
@tbouvet what is your opinion on that? This will affect `docker_swarm` if it starts using `AnsibleDockerSwarmClient` instead of `AnsibleDockerClient`.

I am ok with that proposal, this way we will have one conditional in the constructor and less code in modules checking if we run module on Swarm manager

Reply:
You could also add a function to `AnsibleDockerSwarmClient` which verifies this, and call that function from this constructor. Then it's easy to add (with almost no code repetition), and `docker_swarm` can only call it if `state` is not `join`.

Reply:
I didn't mean the constructor of `AnsibleDockerSwarmClient`, but the one of a deriving class (resp. in `docker_node_facts`'s main() function, since the `docker_node_fact` module uses `AnsibleDockerSwarmClient` directly).

Reply:
I am not sure if that's what you recommended:

In `AnsibleDockerSwarmClient` class new method:
```
    def fail_task_if_not_swarm_manager(self):
        if not self.check_if_swarm_manager():
            self.client.fail(msg="This node is not a manager.")
```

In `docker_node_facts` `main()` before getting facts new call:
```
    client.fail_task_if_not_swarm_manager()

    node = get_node_facts(client)
```

In `docker_node`
```
    client.fail_task_if_not_swarm_manager()

    SwarmNodeManager(client, results)
```

Alternatively, we can in the `AnsibleDockerSwarmClient` add this to constructor and make an exceptions for `state` `join` and `leave` as parameters are available for this class. We can even put all logic, not just checking about swarm manager, based on parameters provided to `AnsibleDockerSwarmClient` in constructor.

Reply:
In `docker_node`, I would call that function from `SwarmNodeManager`'s constructor. Otherwise that looks fine.

I would not have any logic working with `state` option values in `AnsibleDockerSwarmClient`; the client should not depend on how the module's `state` option is designed.